### DISCUSSION TITLE
Absurd patterns

### DISCUSSION THREAD

**Discussion by JakobBruenker** [ghc-proposals/ghc-proposals#423] [Ideas]
The discussion on #302, starting here https://github.com/ghc-proposals/ghc-proposals/pull/302#issuecomment-831342880, has recently been focused on the possibility of introducing absurd patterns. In lieu of a full proposal and in an effort to have a proper place to hold that discussion, I'm creating this issue.

Using `_|_` as tentative syntax to denote an absurd pattern, the idea is that it can appear anywhere where patterns can appear, and signifies that there is no possible non-bottom value that it could match. The compiler would complain if this is not the case. For example:

```haskell
-- currently:
fromBoolVoidAndInt :: (Bool, Void) -> Int -> String
fromBoolVoidAndInt (b, v) x -> case v of {}

-- with absurd pattern:
fromBoolVoidAndInt :: (Bool, Void) -> Int -> String
fromBoolVoidAndInt (b, _|_) x
```

Since an absurd pattern can never be successfully matched, no right-hand side is needed for this equation.

There are a few open questions, though:
 - If an absurd pattern tries to match a bottom value, is the resulting bottom value the matched value or an error about a pattern match failure (a bang in front of the absurd pattern could change this)
 - How exactly should this interact with GADTs?

Feel free to make this issue obsolete by creating a proper proposal.

**Comment by JakobBruenker** [comment]
Btw, @goldfirere, in #302 you say

> If that evaluation diverges, then the pattern-match diverges for the same reason. If that evaluation does not diverge, then the pattern-match diverges with "absurd absurd pattern" or similar.

Is it possible for an evaluation to not diverge here? I would expect that in a correct implementation, the compiler doesn't let you write an absurd pattern unless the only possibility is bottom.

**Comment by cgibbard** [comment]
From https://github.com/ghc-proposals/ghc-proposals/pull/302#issuecomment-833703285
> @cgibbard Hm, it seems fine to me to simply always treat an absurd pattern as strict, and show the error message of the bottom value that was passed as argument. It seems like a simpler solution than distinguishing between these cases, and consistent with how other patterns work, where a bang doesn't change the behavior (unless its a variable or wildcard pattern).

In many cases, that's true, but if the impossible argument was constructed by an accident of recursion, then you might be turning a message with a hint about why the program failed into an infinite loop.

It's yet again a bit of a digression, but it feels as though sometimes it would be nice if GHC could, in the event of an error, collect up all the errors produced by things that it was already sufficiently obvious were going to be forced (e.g. you don't need to do any recursion to get to the point of needing to evaluate them, or they're things involved in the same pattern match), and somehow collect all the errors together, if there happened to be more than one. So in this case, it could combine the pattern match failure from the absurd pattern, and an error from the bottom argument, supposing that it evaluates to one promptly enough. It's probably quite a convoluted change to try to make for such a slight improvement though.

One approach would be for absurd patterns to compile to code that produces an error something like:

```
oops _|_
==>
oops b = error ("Absurd pattern matched in oops: " ++ seq b "")
```

So you get an error which causes another error when printed (or perhaps the infinite loop...).

This might be too cute for its own good, and would surely be frustrating to anyone who actually wanted to catch and log the exception though (but such activity is uncommon with errors thrown by evaluation).

**Comment by JakobBruenker** [comment]
Hm, fair, I suppose the difference to other kinds of patterns is that for them, you'd need to evaluate infinite loops to find out whether they match, whereas for absurd patterns, you know beforehand that they won't match.

**Comment by goldfirere** [comment]
Given that @JakobBruenker both started this Issue and recommended that we try out Discussions, and that the second word in the OP is "discussion", I will convert this to a Discussion. Let's see what happens!

**Comment by phadej** [comment]
I often enough wished to have something like this, but I think that absurd *pattern* is not-as-convenient as marking the RHS as non-reachable. As @JakobBruenker mentions GADTs are problematic.

Simplest example is

```haskell
{-# OPTIONS_GHC -Wall #-}
{-# LANGUAGE GADTs #-}

data Tag a where
    IsInt  :: Tag Int
    IsBool :: Tag Bool

foo :: Tag Int -> ()
foo IsInt = ()
```

Sometimes I wish to be able to write more verbosely as

```haskell
foo :: Tag Int -> ()
foo IsInt = ()
foo IsBool = {-# INACCESSIBLE #-} error "x"
```

real examples are usually something more complicated where having an extra case and explicitly marking it as inaccessible would be helpful (for the reader).

> If an absurd pattern tries to match a bottom value, is the resulting bottom value the matched value or an error about a pattern match failure (a bang in front of the absurd pattern could change this)

I'm quite sure that the error contained in the bottom value is more helpful. Consider `-fdefer-type-errors` which may hide the type error.
I.e. absurdity should be verified by forcing the bottom value.

---

I also suggest that if this feature is proposed, it should deprecate `EmptyCase`, and rather we would write:

```haskell
absurd :: Void -> a
absurd x = case x of
    !_ -> {-# INACCESSIBLE #-}

-- or

absurd :: Void -> a
absurd x = case x of
    _|_ -- this is weird, this is why I don't like it.

-- instead of current
absurd :: Void -> a
absurd x = case x of {}
```

(of course we could write `absurd !_ = {-# INACCESSIBLE #-}` but I wanted to discuss `case`).

This is interesting case (sorry for the pun), as `EmptyCase` evaluates the scrutinee even there is no pattern requiring that - inconsistency in semantics. But `case x of { !_ -> {-# INACCESSIBLE #-} }` fixes that!

---

`{-# INACCESSIBLE #-}` is just some (bad) syntax. Maybe this would be a case for modifiers, if we really would like to avoid introducing new keywords:

```haskell
absurd !_ = %Inaccessible ()
```

**Reply by treeowl** [reply]
I'm very much opposed to deprecating `EmptyCase`, whatever else we do or don't do. It's perfectly sensible syntax. Also, your proposed syntax is horribly verbose. We really just want to be able to make it visually clear that we covered the case—anything further can be in a comment rather than code. Some other languages use special syntax for this. Another option is to use a magical `Inaccessible` identifier. `inaccessible` would have any type, with any runtime representation, but would only be permitted in an inaccessible branch. Then the dead code checker could suppress its warning when the result is `inaccessible`.

**Reply by phadej** [reply]
I argue that it's not sensible. It adds a bang which isn't there in the source. It's inconsistent.

I'm not attached to any of my syntactical proposals, my idea is to have some kind of special RHS, what it could be is up to bikeshedding.

**Comment by treeowl** [comment]
The inconsistency could be resolved by making case unconditionally strict
in its scrutinee, as it is in Core.

**Reply by JakobBruenker** [reply]
You could do that, though it would be a breaking change, since currently `case undefined of x -> ()` doesn't blow up. (Though it does make me wonder how often a case expression is used whose only pattern is a variable or wildcard, I can't think of a sensible use for it at the moment. If it's not used being a breaking change wouldn't matter much, I suppose.)

**Reply by JakobBruenker** [reply]
Though at that point you have a different kind of inconsistency, in that `foo x = x` would behave differently than `foo x = case x of y -> y`. (i.e., variable and wildcard patterns would be strict in `case` but not in functions)

**Reply by phadej** [reply]
Core `case` cannot have nested patterns, so there it makes sense it evaluates. In Haskell that won't make sense:

My proposal is to be able to write:

```haskell
case pairOfIntAndVoid of
   (_, !_) -> inaccessible
```

where now you have to write

```haskell
case pairOfIntAndVoid of
   (_, v) -> case v of {} -- and this implicitly contains the bang from above
```

Prior to `EmptyCase` we wrote

```haskell
case pairOfIntAndVoid of
   (_, v) -> error "impossible"
```

or maybe more correctly

```haskell
case pairOfIntAndVoid of
   (_, v) -> v `seq` error "impossible"
   
-- or

case pairOfIntAndVoid of
   (_, !v) -> error "impossible"
```

See how bang is naturally there. We had to write `error "something"` to make compiler happy, but that RHS is unreachable!

**Comment by treeowl** [comment]
Why is that inconsistent?

**Reply by JakobBruenker** [reply]
Because variable and wildcard patterns then appear to induce strictness in `case` expressions, but not as patterns in a function equation. I would agree that this is a lesser inconsistency than empty case being strict while other case expressions arent'.

Though in light of #302, I'm a bit wary of making case strict in general, since the next step on that path would be making the constructs proposed there strict as well, and I'd like to be able to say things like

```haskell
f :: Maybe [a] -> a -> [a]
f = \mcase
  (Just xs) a -> a : xs
  Nothing a -> [a]
```
 without being strict in `a`.

**Comment by treeowl** [comment]
I think you may have convinced me.

**Comment by Ericson2314** [comment]
Yeah I want a thing for GADTs too. I made an `@`-pattern strawman syntax in the other proposal for that. I guess I'll copy of here when I'm not on my phone.

**Comment by Ericson2314** [comment]
Should we come up with some concrete syntax? Because I want this to work for GADTs and plain empty case, I am thinking it should be
```
alt ::= pat -> expr | pat 'toBeBikeshedKeyword'
```
Note that this is absurd *alternatives* not absurd *patterns*. We aren't deciding which part of the possibly-nested pattern is to "blame".

In particular, we could have
```haskell
data Void
case v :: Void of
  !_ toBeBikeshedKeyword
```
```haskell
case v :: Int :~: Bool of
  Refl toBeBikeshedKeyword
```
```haskell
case v :: Int :~: Bool of
  !_ toBeBikeshedKeyword
```
```haskell
case v :: Identity Void of
  (Identity !_) toBeBikeshedKeyword
```
```haskell
case v :: Identity (Int :~: Bool) of
  (Identity Refl) toBeBikeshedKeyword
```
```haskell
case v :: Identity (Int :~: Bool) of
  (Identity !_) toBeBikeshedKeyword
```
In particular `!_` is nice to be concise, but users are free to *not* use wildcards too to get better errors if a new variant (that perhaps might not be absurd in these patterns) is added to the datatype.

This would allow me to resurrect the spirit of `-XNoIncomplete` by writing a proposal which merely required no missing fields and no missing variants. That would in particular *not* specify what absurd variants are valid, which mean we can have an air-tight spec of the new `-XNoIncomplete` without tackling the really hard task of declaratively specifying the current behavior of the pattern match exhaustiveness checker.

**Comment by treeowl** [comment]
@phadej suggested

```haskell
case pairOfIntAndVoid of
   (_, !_) -> inaccessible
```

Notably, this would not seem to require any syntax changes whatsoever. `inaccessible` would just be a magical identifier exported from `GHC.Magic`, and allowed only as the complete right-hand side of an impossible match.

**Reply by Ericson2314** [reply]
Hmm OK. I feel silly for not scrolling up.

I am not sure about magic identiferi, but there is also condional keyword hacks.

**Comment by Ericson2314** [comment]
Wait.....could we use `unsatisfiable`?!?! We already have some "type error in dead code are OK logic'.

CC @adamgundry

**Reply by Ericson2314** [reply]
OTOH, perhaps this special *deserves* syntax because this is supposed to be the "root" *ex falso quodlibet* construct the others turn into. To use `unsatisfiable` would imply there is some other unwritten more fundamental mechanism.

Vaguely relatedly, I hope we get a better empty-case-concept in Core so that someday if we do have a proper specified exhaustiveness checker, it can be a tactic that emit proofs which core-lint can easily check. The user-written absurd pattern thing would essentially be call to the tactic to put in the proof. But for really simple empty case vs gadt situations, hardly proof is needed, and the surface level construct goes straight to the core one.

**Comment by Ericson2314** [comment]
Yes, per https://github.com/ghc-proposals/ghc-proposals/discussions/423#discussioncomment-1556150 I think this deserves dedicated syntax. Anyone have ideas for that? I might like to write a proposal for it soon, say after #448 is wrapped up (for sake of not having too many things in flight at once).

**Comment by JakobBruenker** [comment]
See https://github.com/ghc-proposals/ghc-proposals/pull/585#issuecomment-1490372899 as a possible solution, though I haven't thought about it too deeply yet
